activesupport/lib/active_support/callbacks.rb

```diff
@@ -372,56 +372,140 @@ def conditions_lambdas
 
       # A future invocation of user-supplied code (either as a callback,
       # or a condition filter).
-      class CallTemplate # :nodoc:
-        def initialize(target, method, arguments, block)
-          @override_target = target
-          @method_name = method
-          @arguments = arguments
-          @override_block = block
+      module CallTemplate # :nodoc:
+        class MethodCall
+          def initialize(method)
+            @method_name = method
+          end
+
+          def expand(target, value, block)
+            [target, block, @method_name]
+          end
+
+          def make_lambda
+            lambda do |target, value, &block|
+              target.send(@method_name, &block)
+            end
+          end
+
+          def inverted_lambda
+            lambda do |target, value, &block|
+              !target.send(@method_name, &block)
+            end
+          end
         end
 
-        # Return the parts needed to make this call, with the given
-        # input values.
-        #
-        # Returns an array of the form:
-        #
-        #   [target, block, method, *arguments]
-        #
-        # This array can be used as such:
-        #
-        #   target.send(method, *arguments, &block)
-        #
-        # The actual invocation is left up to the caller to minimize
-        # call stack pollution.
```

Comment:
... given we're no longer doing the generic map inside expand, though, is there any value to removing the splat from its return value, and making it return exactly 4 elements, the last of which is an `arguments` array?

Reply:
I have to admit I don't have a full grasp of the callbacks implementation, but by digging more I'd like to find a way to get rid of `expand`.

We definitely should keep the documentation, I'm just not sure to put it since there's no longer a base type, I migth just move it to the module documentation.

Reply:
I don't think we can do much to drop expand without degrading the backtrace (which I'd really prefer to avoid -- we have callbacks everywhere, and they're noisy enough as is), or enabling tailcalls (which I hadn't realised are now effectively unavailable https://bugs.ruby-lang.org/issues/15303).

("not much" > "nothing", because we _could_ replace it with four separate methods, which would skip the array allocation... or, again as there's only one caller, some alternative to effectively inline it as a `case` statement, say)

Reply:
> we could replace it with four separate methods, which would skip the array allocation...

Yeah that's something I had in mind but haven't had time to explore yet (I had to drop this PR for now to deal with Ruby 3.0.3 / bootsnap, I'll clean it up tomorrow based on your feedback).

Either way if this PR is good enough of a step for you, I'd rather explore other improvements in followups (if I can actually find other improvements).

Reply:
:+1:

In many places I might be dubious about the benefit vs the increase in code (at least to the point of wanting to look at more holistic benchmarks, to prove the callback-wrapped action itself doesn't automatically drown out the improvement here), but this is exactly the sort of low-level Active Support code where it fairly-automatically makes sense to make the trade-off. (As is consistent with the existing state of this file.)

Reply:
> at least to the point of wanting to look at more holistic benchmarks, to prove the callback-wrapped action itself doesn't automatically drown out the improvement here

Yeah, I'm not sure how best to benchmark this code.

For the record, what prompted me to look at this is @jhawthorn's recent ruby PR https://github.com/ruby/ruby/pull/5146 which massively speeds up Rails. And we were talking about that `instance_exec` issue since a few months, and we assumed it was mostly from callbacks, so I suspect they're quite the hotspot.

I'll see if this improvement is visible on rails-bench, but I suspect it might be a bit too small.

Reply:
> we could replace it with four separate methods, which would skip the array allocation...

So for the record I just tried this, and it's marginally slower :/. Seems like the extra method calls happen cost slightly more than the extra allocation.